### PR TITLE
Preload BPF loader

### PR DESCRIPTION
**Pull request by jackcmay**
This is an intermediate change to allow folks use the dynamic BPF loader.  A more long term solution is being discussed in #1565

### CODE REVIEW

**Comment by jackcmay** [comment]
https://github.com/solana-labs/solana/issues/1565

File: src/bank.rs [hunk 1]
```diff
@@ -195,6 +196,14 @@ impl Bank {
             let account = accounts.entry(deposit.to).or_insert_with(Account::default);
             Self::apply_payment(deposit, account);
         }
+        {
```
Comment (mvines) [inline src/bank.rs]: Can you move this block into a new `bpf_loader` fn.  Looks like it'll just require a `bank` reference.  This should limit the bank.rs touching

File: src/bpf_loader.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+//! BPF loader
+use native_loader;
+use solana_program_interface::account::Account;
+use solana_program_interface::pubkey::Pubkey;
+
+pub const BPF_LOADER_PROGRAM_ID: [u8; 32] = [6u8; 32];
+pub const BPF_LOADER_NAME: &str = "bpf_loader";
+
+pub fn id() -> Pubkey {
```
Comment (mvines) [inline src/bpf_loader.rs]: nit: `program_id()` seems nicer.   the name `id()` is too generic.  But I see `native_loader::id()` so we're already confused.   I sorta just want to delete this function entirely, it barely adds value

File: tests/programs.rs [hunk 3]
```diff
@@ -260,6 +271,39 @@ fn test_program_lua_move_funds() {
...
+        elf::File::open_path(&create_bpf_path("noop_c"))
+            .unwrap()
+            .get_section(PLATFORM_SECTION_C)
+            .unwrap()
+            .data
+            .clone(),
+    );
+
+    // Call user program
+
```
Comment (mvines) [inline tests/programs.rs]: nit: remove whitespace

File: src/bpf_loader.rs (hunk 2 above)
Comment (jackcmay) [reply src/bpf_loader.rs]: I agree, just following convention with the rest of the programs.  If we change it should change it consistently across all the programs.  Mind if I leave it consistent with other programs and we can change them all in another pr?

**Comment by jackcmay** [comment]
Added change to fullnode so that bpf loader is picked up

File: src/bank.rs [hunk 4]
```diff
@@ -187,6 +188,24 @@ impl Default for Bank {
...
+        let bank = Self::default();
+
+        {
+            // Preload Bpf Loader account
+            let mut accounts = bank.accounts.write().unwrap();
+            let mut account = accounts
+                .entry(bpf_loader::id())
+                .or_insert_with(Account::default);
+            bpf_loader::populate_account(&mut account);
+            println!("BPF Loader Built-in!");
```
Comment (mvines) [inline src/bank.rs]: println! 🔪

File: src/bank.rs [hunk 5]
```diff
@@ -187,6 +188,24 @@ impl Default for Bank {
 }
 
 impl Bank {
+    /// Create an Bank with built-in programs.
+    pub fn new_default() -> Self {
```
Comment (mvines) [inline src/bank.rs]: how about `new_with_builtin_programs()`

File: src/bank.rs [hunk 6]
```diff
@@ -187,6 +188,22 @@ impl Default for Bank {
...
+                .or_insert_with(Account::default);
+            bpf_loader::populate_account(&mut account);
+        }
+
+        bank
+    }
+
     /// Create an Bank using a deposit.
     pub fn new_from_deposit(deposit: &Payment) -> Self {
         let bank = Self::default();
```
Comment (mvines) [inline src/bank.rs]: oh can we use `Self::new_with_builtin_programs()` here now, to de-dup the "Preload Bpf Loader account" block?

**Comment by garious** [comment]
Why isn't this merged?

**Comment by jackcmay** [comment]
One final change to make

**Comment by jackcmay** [comment]
3 Consecutive CI failures due to `test_leader_restart_validator_start_from_old_ledger`!

**Pull request merged**